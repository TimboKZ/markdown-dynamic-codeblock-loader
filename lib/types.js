"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChunkType = void 0;
var ChunkType;
(function (ChunkType) {
    ChunkType["Markdown"] = "markdown";
    ChunkType["Code"] = "code";
})(ChunkType = exports.ChunkType || (exports.ChunkType = {}));
//# sourceMappingURL=types.js.map