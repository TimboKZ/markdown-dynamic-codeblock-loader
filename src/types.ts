export interface LoaderOptions {
    mappings: PathMapping;
}

export interface CodeBlockData {
    start: number;
    end: number;
    language: string;
    jsonConfig: string;
}

export enum ChunkType {
    Markdown = 'markdown',
    Code = 'code',
}

export interface MarkdownChunk {
    type: ChunkType.Markdown;
    contents: string[];
}

export interface CodeChunk {
    type: ChunkType.Code;
    codeBlock: CodeBlockData;
    openingLine: string;
    contents: string[];
    closingLine: string;
}

export type Chunk = MarkdownChunk | CodeChunk;

export type PathMapping = { [name: string]: string };
