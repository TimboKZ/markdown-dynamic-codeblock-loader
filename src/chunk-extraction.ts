import { Chunk, ChunkType } from './types';

const codeBlockStartRegex = /^\s*```([A-Za-z0-9\-_]+)?\s*({\s*.*?\s*})\s*$/;
const codeBlockEndRegex = /^\s*```\s*$/;

export interface CodeblockData {
    start: number;
    end: number;
    language: string;
    jsonConfig: string;
}

/**
 * Scans lines of a Markdown files and detect code blocks, extract info like
 * first/last lines and syntax name.
 */
export function detectCodeBlocks(lines: string[]): CodeblockData[] {
    let insideCodeBlock = false;
    let blockStartIndex: number = -1;
    let blockLanguage: string = '';
    let blockJsonConfig: string = '';
    const codeBlocks: CodeblockData[] = [];

    for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
        const line = lines[lineIndex];

        const codeStartMatch = line.match(codeBlockStartRegex);
        if (codeStartMatch) {
            insideCodeBlock = true;
            blockStartIndex = lineIndex;
            [, blockLanguage, blockJsonConfig] = codeStartMatch;
            continue;
        }

        const codeEndMatch = line.match(codeBlockEndRegex);
        if (codeEndMatch) {
            if (insideCodeBlock) {
                codeBlocks.push({
                    start: blockStartIndex,
                    end: lineIndex,
                    language: blockLanguage.trim(),
                    jsonConfig: blockJsonConfig.trim(),
                });
            }
            insideCodeBlock = false;
        }
    }

    return codeBlocks;
}

/**
 * Scans lines of Markdown files, breaking it into Markdown and Code chunks
 */
export const breakMarkdownIntoChunks = (lines: string[]): Chunk[] => {
    const codeBlockIndices = detectCodeBlocks(lines);
    const markdownChunks: Chunk[] = [];

    let markdownStartIndex = 0;
    let markdownSlice = [];
    for (const block of codeBlockIndices) {
        const { start, end, language, jsonConfig } = block;

        markdownSlice = lines.slice(markdownStartIndex, start);
        if (markdownSlice.length > 0) {
            markdownChunks.push({
                type: ChunkType.Markdown,
                contents: markdownSlice,
            });
        }

        markdownChunks.push({
            type: ChunkType.Code,
            language,
            jsonConfig,
            openingLine: lines[start],
            contents: lines.slice(start + 1, end),
            closingLine: lines[end],
        });

        markdownStartIndex = end + 1;
    }

    markdownSlice = lines.slice(markdownStartIndex, lines.length);
    if (markdownSlice.length > 0) {
        markdownChunks.push({
            type: ChunkType.Markdown,
            contents: markdownSlice,
        });
    }

    return markdownChunks;
};

export const stitchMarkdownChunksIntoMarkdown = (markdownChunks: Chunk[]) => {
    let markdownString = '';
    for (const chunk of markdownChunks) {
        if (chunk.type === ChunkType.Code) {
            markdownString += `${chunk.openingLine}\n`;
            markdownString += `${chunk.contents.join('\n')}\n`;
            markdownString += `${chunk.closingLine}\n`;
        } else {
            markdownString += `${chunk.contents.join('\n')}\n`;
        }
    }
    return markdownString;
};
