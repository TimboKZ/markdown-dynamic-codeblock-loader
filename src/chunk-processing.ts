import fs from 'fs';
import { stringifyRequest } from 'loader-utils';
import path from 'path';
import webpack from 'webpack';
import { extractSymbolLinesFromFile } from './symbol-extraction';

import { CodeChunk, PathMapping } from './types';
import LoaderContext = webpack.loader.LoaderContext;

export const processCodeChunk = (
    loaderContext: LoaderContext,
    chunk: CodeChunk,
    pathMapping: PathMapping
) => {
    const { jsonConfig } = chunk.codeBlock;
    if (!jsonConfig) return;

    let config;
    try {
        config = JSON.parse(jsonConfig);

        if (!config.file || !config.symbol) {
            throw new Error('Config should specify both "file" and "symbol" fields.');
        }

        // Apply path mappings provided through loader options
        let expandedPath = config.file;
        for (const token in pathMapping) {
            if (!pathMapping.hasOwnProperty(token)) continue;
            expandedPath = expandedPath.replace(`<${token}>`, pathMapping[token]);
        }

        const filePath = path.resolve(
            path.dirname(loaderContext.resourcePath),
            expandedPath
        );
        if (!fs.existsSync(filePath)) {
            throw new Error(
                `Provided "file" does not exist. Resolved path: ${filePath}`
            );
        }

        const relativeFilePath = stringifyRequest(loaderContext, filePath);
        loaderContext.addDependency(relativeFilePath);

        chunk.contents = extractSymbolLinesFromFile(filePath, config.symbol);
        return filePath;
    } catch (error) {
        throw new Error(
            `Failed to process the code block on line ${chunk.codeBlock.start + 1}: ` +
                `${error.message}` +
                `\nThe JSON config for the code block was: ${jsonConfig}`
        );
    }
};
