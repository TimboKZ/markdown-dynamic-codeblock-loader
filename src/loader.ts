import { getOptions } from 'loader-utils';
import validateOptions from 'schema-utils';
import webpack from 'webpack';

import {
    breakMarkdownIntoChunks,
    stitchMarkdownChunksIntoMarkdown,
} from './chunk-extraction';
import { processCodeChunk } from './chunk-processing';
import { ChunkType } from './types';
import LoaderContext = webpack.loader.LoaderContext;

const schema = {
    type: 'object',
    properties: {
        mappings: {
            type: 'object',
        },
    },
} as const;

const LinebreakRegex = /\r?\n/;

export function loader(this: LoaderContext, source: string): string {
    const options = getOptions(this);
    validateOptions(schema, options, {
        name: 'Markdown Dynamic Codeblock Loader',
        baseDataPath: 'options',
    });

    const lines = source.split(LinebreakRegex);
    const chunks = breakMarkdownIntoChunks(lines);

    for (const chunk of chunks) {
        if (chunk.type !== ChunkType.Code) continue;

        processCodeChunk(this, chunk);
    }

    const transformedMarkdown = stitchMarkdownChunksIntoMarkdown(chunks);

    return `export default ${JSON.stringify(transformedMarkdown)}`;
}
