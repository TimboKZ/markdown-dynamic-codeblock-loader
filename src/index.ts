import { loader } from './loader';

// eslint-disable-next-line
export default loader;
