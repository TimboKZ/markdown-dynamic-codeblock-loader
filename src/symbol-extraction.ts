import { parse } from '@babel/parser';
import { Statement } from '@babel/types';
import fs from 'fs';
import { Nullable } from 'tsdef';
import { LinebreakRegex } from './loader';

export const extractSymbolLinesFromFile = (
    filePath: string,
    symbol: string
): string[] => {
    const fileSource = fs.readFileSync(filePath).toString();

    const tree = parse(fileSource, {
        sourceType: 'module',
        plugins: ['typescript'],
    });

    // Walk through all nodes in a program
    const programNodes = tree.program.body;
    let symbolLocation: Nullable<SymbolLocation> = null;
    for (const node of programNodes) {
        symbolLocation = tryToFindSymbol(node, symbol);
        if (symbolLocation) break;
    }

    if (!symbolLocation) {
        throw Error(
            `Could not find symbol ${symbol} in the resolved file: ${filePath}`
        );
    }

    const lines = fileSource
        .split(LinebreakRegex)
        .slice(symbolLocation.firstLineIndex, symbolLocation.lastLineIndex + 1);

    return lines;
};

interface SymbolLocation {
    firstLineIndex: number;
    lastLineIndex: number;
}

const tryToFindSymbol = (node: Statement, symbol: string): Nullable<SymbolLocation> => {
    if (node.type === 'ExportNamedDeclaration') {
        const declarationNode = node.declaration;
        if (declarationNode) {
            if (declarationNode.type === 'TSInterfaceDeclaration') {
                if (declarationNode.id.name === symbol) {
                    const loc = node.loc;
                    if (loc && loc.start && loc.end) {
                        return {
                            firstLineIndex: loc.start.line - 1,
                            lastLineIndex: loc.end.line - 1,
                        };
                    }
                }
            }
        }
    }
    return null;
};
