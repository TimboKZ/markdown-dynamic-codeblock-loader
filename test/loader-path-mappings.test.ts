import { describe, expect, it } from '@jest/globals';
import fs from 'fs';
import path from 'path';

import { LoaderOptions } from '../src/types';
import { getTransformedMarkdown } from './compiler';

describe('Path mappings', () => {
    const folderPath = path.resolve(__dirname, 'path-mappings');

    it('Are resolved correctly', async () => {
        const loaderOptions: LoaderOptions = {
            mappings: {
                // Absolute paths!
                'folder-a': path.join(folderPath, 'folder-a'),
                'folder-b': path.join(folderPath, 'folder-b'),
            },
        };

        const inputFilePath = path.join(folderPath, 'valid.path-mappings.input.md');
        const outputFilePath = path.join(folderPath, 'valid.path-mappings.output.md');

        const transformedMarkdown = await getTransformedMarkdown(
            inputFilePath,
            loaderOptions
        );
        fs.writeFileSync(outputFilePath, transformedMarkdown);
        expect(transformedMarkdown).toMatchSnapshot();
    });
});
