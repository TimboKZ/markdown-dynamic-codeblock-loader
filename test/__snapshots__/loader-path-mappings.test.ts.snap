// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path mappings Are resolved correctly 1`] = `
"# Using path mappings in loader options

\`\`\`ts { \\"file\\": \\"<folder-a>/folder-a.ts\\", \\"symbol\\": \\"AaaaAAaaaaAAA\\" }
export interface AaaaAAaaaaAAA {
    AaAAaaaAAaa: string;
}
\`\`\`

\`\`\`ts { \\"file\\": \\"<folder-b>/folder-b.ts\\", \\"symbol\\": \\"BbbbBBbbbbBBB\\" }
export interface BbbbBBbbbbBBB {
    BbBBbbbBBbb: string;
}
\`\`\`

"
`;
